### DISCUSSION TITLE
Encode exception details as json

### DISCUSSION THREAD

**Discussion by giacomofiorindo** [fastapi/fastapi#9096] [Questions]
Hi, 
I was working with FastApi  and Pydantic when I noticed that passing a Pydantic model as detail to a HttpException causes another exception (if the model contains normally unjsonable fields such as datetime):

A simple example:
```
from pydantic import BaseModel
from datetime import date


class Example(BaseModel):
    name: str
   date: date
```
```
example = Example(name='example', date=date.today())
raise HTTPException(
            status_code=400,
            detail=example
        )
```
This would cause a `TypeError: Object of type Example is not JSON serializable`

This could be easily fixed by changing `exception_handlers.py`:
```
async def http_exception_handler(request: Request, exc: HTTPException) -> JSONResponse:
    headers = getattr(exc, "headers", None)
    if headers:
        return JSONResponse(
-             {"detail": exc.detail}, status_code=exc.status_code, headers=headers
+            {"detail": jsonable_encoder(exc.detail)}, status_code=exc.status_code, headers=headers
        )
    else:
-        return JSONResponse({"detail": exc.detail}, status_code=exc.status_code)
+        return JSONResponse({"detail": jsonable_encoder(exc.detail)}, status_code=exc.status_code)
```

Since it might be intentional, I've labelled the issue as a suggestion

P.S. calling json() on the model when raising the exception works too, but the result has ugly formatting and escape characters

**Comment by peterroelants** [comment]
I'm also running into this and [from the documentation](https://fastapi.tiangolo.com/tutorial/handling-errors/#:~:text=converted%20to%20JSON%20as%20the%20parameter%20detail) I thought this was expected behaviour:
> When raising an HTTPException, you can pass any value that can be converted to JSON as the parameter detail, not only str.
You could pass a dict, a list, etc.
They are handled automatically by FastAPI and converted to JSON.

Using `Model.dict()` works, so it's not a major issue.

**Comment by sayoojt** [comment]
I have used jsonable_encoder.

from fastapi.encoders import jsonable_encoder

 print(jsonable_encoder(example))

**Comment by jd-solanki** [comment]
Both `jsonable_encoder` and `.model_dump()` (pydantic v2) is valid for this but support for this out of the box suits the DX of FastAPI & also how greatly FastAPI works with Pydantic.

**Comment by YuriiMotov** [comment]
Test:
```python
from fastapi import FastAPI, HTTPException
from fastapi.testclient import TestClient
from pydantic import BaseModel
from datetime import date


class Example(BaseModel):
    name: str
    date: date


app = FastAPI()


@app.get("/")
async def index():
    example = Example(name="example", date=date.today())
    raise HTTPException(status_code=400, detail=example)


def test_():
    client = TestClient(app)
    resp = client.get("/")  # Raises TypeError: Object of type Example is not JSON serializable
    assert resp.status_code == 200
```
